account_invoice_mode_monthly/models/sale_order.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from datetime import datetime
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import api, models
+from odoo.osv.expression import OR
+
+from odoo.addons.queue_job.job import job
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model
+    def cron_generate_monthly_invoices(self):
+        """Cron called daily to check if monthly invoicing needs to be done."""
+        company_ids = self._company_monthly_invoicing_today()
+        if company_ids:
+            self.generate_monthly_invoices(company_ids)
+
+    @api.model
+    def generate_monthly_invoices(self, companies=None):
+        """Generate monthly invoices for customers who require that mode.
+
+        Invoices will be generated by other jobs split for different customer
+        and different payment term.
+        """
+        if not companies:
+            companies = self.company_id
+        saleorder_groups = self.read_group(
+            [
+                ("invoicing_mode", "=", "monthly"),
+                ("invoice_status", "=", "to invoice"),
+                ("company_id", "in", companies.ids),
+            ],
+            ["partner_invoice_id"],
+            groupby=self._get_groupby_fields_for_monthly_invoicing(),
+            lazy=False,
+        )
+        for saleorder_group in saleorder_groups:
+            saleorder_ids = self.search(saleorder_group["__domain"]).ids
+            self.with_delay()._generate_invoices_by_partner(saleorder_ids)
+        companies.write({"invoicing_mode_monthly_last_execution": datetime.now()})
```

Comment:
why don't you compute and store the next invoice day instead? Your search query will be much more simple I think, no?

Reply:
The query may look complicated but it is covered by tests.
I am not convince that storing the next execution date would make things simpler. One of the reason is the day of invoicing in the month can be set in the configuration.
So now when changing that day the next invoicing day as too be computed but depending if the invoicing day as been done for the current month or not the computation would be different. (This is only one complication I can think of now)